### PR TITLE
Fix NuGetHome resolution on Kudu

### PR DESCRIPTION
**Pull request by joelverhagen**
https://github.com/NuGet/Home/issues/1981

This re-applies a [pre-existing fix](https://github.com/NuGet/NuGet.Client/commit/381b6128d74cd51df82544b936533331452123a1#diff-f83497230965c941a0bf5e2d2ed13deeR151) that was presumably lost during a refactor. Tested by deploying to [an Azure website](http://nugetjver.azurewebsites.net/api/values/NuGetHome).

Before: `NuGetFolderPath.NuGetHome` resolves to `.nuget`
After: `NuGetFolderPath.NuGetHome` resolves to `D:\local\UserProfile\.nuget`

@emgarten @yishaigalatzer @glennc

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs [hunk 1]
```diff
@@ -25,6 +25,13 @@ public static string GetFolderPath(NuGetFolderPath folder)
                         "NuGet");
                 case NuGetFolderPath.NuGetHome:
                     var userProfile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
+#if !DNXCORE50
```
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: On when building for `DNXCORE50`, `Environment.GetFolderPath` calls into the code below, which already calls `GetHome`, so no need to call it again.

Comment (emgarten) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: This is needed to keep GetHome from being called twice under DNX?

**Comment by emgarten** [comment]
:shipit:

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Spoke offline, yes that is the intent.

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs [hunk 2]
```diff
@@ -44,15 +51,42 @@ public static string GetFolderPath(NuGetFolderPath folder)
...
         }
 
-#if DNXCORE50
-        private static class Environment
+        private static string GetHome()
         {
-            public static string NewLine { get; } = System.Environment.NewLine;
+            if (RuntimeEnvironmentHelper.IsWindows)
+            {
+                return FirstNonEmpty(
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Please simplify this. No need for this smarts and extra indirections

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs [hunk 3]
```diff
@@ -44,15 +51,42 @@ public static string GetFolderPath(NuGetFolderPath folder)
...
+            else
+            {
+                return GetEnvironmentVariable("HOME");
+            }
+        }
 
-            public static string GetEnvironmentVariable(string variable)
+        public static string GetEnvironmentVariable(string variable)
+        {
+            return System.Environment.GetEnvironmentVariable(variable);
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Can you define the using statement on top?

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs [hunk 4]
```diff
@@ -44,15 +51,42 @@ public static string GetFolderPath(NuGetFolderPath folder)
...
+                    return value;
+                }
             }
+            return null;
+        }
+
+#if DNXCORE50
+        private static class Environment
+        {
+            public static string NewLine { get; } = System.Environment.NewLine;
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: What is the point of this cache?

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs [hunk 5]
```diff
@@ -44,15 +51,42 @@ public static string GetFolderPath(NuGetFolderPath folder)
...
+                if (!string.IsNullOrEmpty(value))
+                {
+                    return value;
+                }
             }
+            return null;
+        }
+
+#if DNXCORE50
+        private static class Environment
```
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Rename to a meaningful name and thus prevent the namespace collision

Comment (emgarten) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: I believe the collision is by design here. The special folder classes will not exist on coreclr and it will call into this instead. (not 100% sure)

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 3 above)
Comment (emgarten) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: This needs to be specific to the System.Environment class for the reason below

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 5 above)
Comment (yishaigalatzer) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Which is why in want of simplified. What we have is a hack, makes the code hard to follow and like you said yourself "not quite sure why it is there"

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 2 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Done.

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 3 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Refactored.

File: src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Common/NuGetEnvironment.cs]: Unused. Removed.

**Comment by emgarten** [comment]
:shipit:
